common/persistence/class.KeyValuePersistence.php

```diff
@@ -14,54 +14,421 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  *
- * Copyright (c) 2013 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ * Copyright (c) 2013-2017 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
  *
  * @author Lionel Lecaque  <lionel@taotesting.com>
+ * @author Camille Moyon  <camille@taotesting.com>
  * @license GPLv2
  * @package generis
  
  *
  */
 class common_persistence_KeyValuePersistence extends common_persistence_Persistence
 {
-    
+    const MAX_VALUE_SIZE = 'max_value_size';
+    const MAP_IDENTIFIER = 'map_identifier';
+
+    const START_MAP_DELIMITER = 'start_map_delimiter';
+    const END_MAP_DELIMITER = 'end_map_delimiter';
+    const MAPPED_KEY_SEPARATOR = '###';
+    const LEVEL_SEPARATOR = '-';
+
+    const DEFAULT_MAP_IDENTIFIER = '<<<<mapped>>>>';
+    const DEFAULT_START_MAP_DELIMITER = '<<<<mappedKey>>>>';
+    const DEFAULT_END_MAP_DELIMITER = '<<<</mappedKey>>>>';
+
+
+    /**
+     * @var int The maximum size allowed for the value
+     */
+    protected $size = false;
+
+    /**
+     * Set a $key with a $value
+     * If $value is too large, it is split into multiple $mappedKey.
+     * These new keys are serialized and stored into actual $key
+     *
+     * @param $key
+     * @param $value
+     * @param null $ttl
+     * @return bool
+     * @throws common_Exception If size is misconfigured
+     */
     public function set($key, $value, $ttl = null)
     {
+        if ($this->isLarge($value)) {
+            common_Logger::t('Large value detected into KeyValue persistence. Splitting value for key : ' . $key);
+            $value = $this->setLargeValue($key, $value, 0, true, true, $ttl);
+        }
         return $this->getDriver()->set($key, $value, $ttl);
     }
-    
+
     /**
-     * Returns the value stored for the identifier $key
-     * or FALSE if nothing found
-     * 
+     * Get $key from driver. If $key is split, all mapped values are retrieved and join to restore original value
+     *
      * @param string $key
+     * @return bool|int|null|string
      */
-    public function get($key) {
-        return $this->getDriver()->get($key);
-    }
-    
-    public function exists($key) {
-        return $this->getDriver()->exists($key);
-    }
-    
-    public function del($key) {
-        return $this->getDriver()->del($key);
+    public function get($key)
+    {
+        $value = $this->getDriver()->get($key);
+        if ($this->isSplit($value)) {
+            common_Logger::t('Large value detected into KeyValue persistence. Joining value for key : ' . $key);
+            $value = $this->join($key, $value);
+        }
+        return $value;
     }
 
-    public function incr($key) {
```

Comment:
`incr` and `decr` are now missing from class KeyValuePersistence which is causing trouble.